Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -64,45 +63,212 @@ extension PluginTarget {
         }
         catch {
             return callbackQueue.async { completion(.failure(PluginEvaluationError.couldNotCreateOuputDirectory(path: outputDirectory, underlyingError: error))) }
         }
 
-        // Create the input context to send to the plugin.
-        var serializer = PluginScriptRunnerInputSerializer(fileSystem: fileSystem, buildEnvironment: buildEnvironment)
-        let inputStruct: PluginScriptRunnerInput
+        // Serialize the plugin action to send as the initial message.
+        let initialMessage: Data
         do {
-            inputStruct = try serializer.makePluginScriptRunnerInput(
-                rootPackage: package,
-                pluginWorkDir: outputDirectory,
-                toolSearchDirs: toolSearchDirectories,
-                toolNamesToPaths: toolNamesToPaths,
-                pluginAction: action)
+            var serializer = PluginContextSerializer(fileSystem: fileSystem , buildEnvironment: buildEnvironment)
+            let rootPackageId = try serializer.serialize(package: package)
+            let pluginWorkDirId = try serializer.serialize(path: outputDirectory)
+            let toolSearchDirIds = try toolSearchDirectories.map{ try serializer.serialize(path: $0) }
+            let toolNamesToPathIds = try toolNamesToPaths.mapValues{ try serializer.serialize(path: $0) }
+            let wireInput = WireInput(
+                paths: serializer.paths,
+                targets: serializer.targets,
+                products: serializer.products,
+                packages: serializer.packages,
+                pluginWorkDirId: pluginWorkDirId,
+                toolSearchDirIds: toolSearchDirIds,
+                toolNamesToPathIds: toolNamesToPathIds)
+            let actionMessage: HostToPluginMessage
+            switch action {
+                
+            case .createBuildToolCommands(let target):
+                guard let targetId = try serializer.serialize(target: target) else {
+                    throw StringError("unexpectedly was unable to serialize target \(target)")
+                }
+                actionMessage = .createBuildToolCommands(
+                    context: wireInput,
+                    rootPackageId: rootPackageId,
+                    targetId: targetId)
+            
+            case .performCommand(let arguments):
+                actionMessage = .performPackageUserCommand(
+                    context: wireInput,
+                    rootPackageId: rootPackageId,
+                    arguments: arguments)
+            }
+            
+            initialMessage = try actionMessage.toData()
         }
         catch {
             return callbackQueue.async { completion(.failure(PluginEvaluationError.couldNotSerializePluginInput(underlyingError: error))) }
         }
+        
+        // Handle messages and output from the plugin.
+        class ScriptRunnerDelegate: PluginScriptRunnerDelegate {
+            let callbackQueue: DispatchQueue
```

Comment:
is this the delegate queue or for some other callback?

Reply:
It's used for both the delegate and the completion block, and it was the name used for this before this PR as well.  Possibly it should be renamed, but `delegateQueue` might suggest it is only used for the delegate.

So a typical client would pass in a serial queue and thus get a sequence of zero or more non-overlapping delegate calls and then finally the completion handler on the same queue.

I've always been a bit on the fence about whether the completion block should be a `didFinish` kind of delegate call, but it's convenient to have it separate if the delegate is optional and there are cases where the completion handler is needed but a delegate isn't.

Reply:
BTW, the only reason the delegate has the queue as a property here is for the dispatch queue precondition check.  Perhaps we should drop that and only do that in unit tests.

Reply:
Removed this delegate actually, and also added comments for the purpose of the properties.